File: src/main/cpp/lib/BooksOrder.cpp
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

#include "client/BooksOrder.h"
#include "MainRepository.h"
#include "storage/Repository.h"
#include "purchase/PurchasedBook.h"

using namespace std;
using namespace purchase;
using namespace domain::book;

namespace client
{
	BooksOrder::BooksOrder(const Client& client) : client_(client)
	{
	}

	void BooksOrder::addBook(const shared_ptr<Book>& book, int quantity)
	{
		booksInBasket_[book] += quantity;
	}

	std::shared_ptr<Invoice> BooksOrder::checkOut()
	{
		std::shared_ptr<Invoice> invoice(std::make_shared<Invoice>(client_.getName(), client_.getCountry()));

		for (auto books_in_basket : booksInBasket_)
		{
			invoice->addPurchasedBook(std::make_shared<PurchasedBook>(books_in_basket.first, books_in_basket.second));
		}

		main_repository::configuredRepository()->addInvoice(invoice);

		return invoice;
	}

	int BooksOrder::getQuantityOf(const shared_ptr<Book>& book) const
	{
		if (booksInBasket_.find(book) == booksInBasket_.end())
			return 0;

		return booksInBasket_.at(book);
	}

}

File: src/main/csharp/Application/Client/BooksOrder.cs
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/licenses/MIT
 *
 *******************************************************************************/

using System.Collections.Generic;
using Application.Domain.Book;
using Application.Purchase;

namespace Application.Client
{
   public sealed class BooksOrder : IOrder
    {
        private readonly Dictionary<IBook, int> _booksInBasket;
        private readonly Client _client;

        public BooksOrder(Client client)
        {
            _client = client;
            _booksInBasket = new Dictionary<IBook, int>();
        }

        public void AddBook(IBook book, int quantity)
        {
            var existingQuanty =  _booksInBasket.GetValueOrDefault(book,0);
            _booksInBasket[book] = existingQuanty + quantity;
        }

        public Invoice CheckOut()
        {
            var invoice = new Invoice(_client.Name, _client.Country);
            
            foreach (var (book, quantity) in _booksInBasket)
                invoice.AddPurchasedBook(new PurchasedBook(book, quantity));

            MainRepository.ConfiguredRepository.AddInvoice(invoice);
            return invoice;
        }

        public int GetQuantityOf(IBook book) => _booksInBasket.GetValueOrDefault(book, 0);
    }
}
